Add exponential weights

ararslan opened the pull request:
This is almost entirely code written by @rofinn. I'm just submitting it here with a couple of slight modifications.

joshday commented on docs/src/weights.md:
```diff
@@ -41,6 +41,16 @@ w = ProbabilityWeights([0.2, 0.1, 0.3])
+### `ExponentialWeights`
+
+Exponential weights are a common form of temporal weights which assign exponentially greater
+weight to past observations.
```
Exponential weights are essentially an easier-to-implement moving window calculation, so shouldn't this be greater weight to recent observations?

rofinn replied:
Yes, this should probably be "Exponential weights are a common form of temporal weights which assigns exponentially decreasing weights to past observations"? On a similar note, does having the order of the values increasing rather than decreasing by default make sense (I typically have more recent observations at the end of an array or dataframe)?

joshday replied:
> I typically have more recent observations at the end of an array or dataframe

I'm guessing that's the more common case, so I think increasing weights is the right default.

joshday commented on src/weights.jl:
```diff
@@ -193,9 +195,87 @@ pweights(vs::RealArray) = ProbabilityWeights(vec(vs))
+"""
+    ExponentialWeights(vs)
+
+Construct an `ExponentialWeights` vector with weight values `vs`, which must sum to 1.
```
Is the sum to 1 necessary?

rofinn replied:
Not strictly necessary, but the `varcorrection` optimization is kind of a nice property.

joshday replied:
I'll admit I'm not entirely sure how `varcorrection` gets used, but normalizing the weights is unexpected (as far as I understand how people use exponential weights).

ararslan commented on src/weights.jl:
```diff
@@ -193,9 +195,87 @@ pweights(vs::RealArray) = ProbabilityWeights(vec(vs))
+
+Construct an [`ExponentialWeights`](@ref) vector using the given array.
+"""
+eweights(v::RealVector) = ExponentialWeights(v)
```
I agree. @rofinn, what was the use case for this?

rofinn replied:
Hmmm, I think it was largely used as a shorthand when constructor eweights in the same way we used in our legacy MATLAB system, but that's probably better handled by the `ExponentialWeights` constructor (from an API standpoint).

ararslan commented on src/weights.jl:
```diff
@@ -193,9 +195,87 @@ pweights(vs::RealArray) = ProbabilityWeights(vec(vs))
+function eweights(n::Integer, λ::Real)
+    n > 0 || throw(ArgumentError("cannot construct exponential weights of length < 1"))
+    0 < λ <= 1 || throw(ArgumentError("smoothing factor must be between 0 and 1"))
+    w0 = map(i -> λ * (1 - λ)^(1 - i), 1:n)
```
`setindex!` is defined for all `AbstractWeights`, but if we don't store the values then we can't `setindex!`. I still think weights should be immutable (though I know that discussion has occurred already), but as long as the other weight types are, `ExponentialWeights` should be as well, IMO.

nalimilan replied:
Well, the difference is that other types can be made mutable for free, while for exponential weights this requires allocating a vector. So I think the inconsistency is worth it.

ararslan replied:
`ExponentialWeights` was literally Rory's use case for mutable weights though :laughing:

nalimilan replied:
How so?

ararslan replied:
See e.g. https://github.com/JuliaStats/StatsBase.jl/issues/397#issuecomment-405793482

nalimilan replied:
If reversing exponential weights is a common need, better add an argument and field to do that?

ararslan replied:
@rofinn, see above--thoughts?

rofinn replied:
FWIW, I don't think we should compute the weights on the fly as this seems more appropriate for an iterable rather than a subtype of `AbstractArray`.

nalimilan replied:
Why? Ranges compute values on the fly on they are `AbstractArray`. As long as it's efficient, I don't see the problem.

rofinn replied:
Yes, but I'm not likely to want to multiple a data matrix by a range.

nalimilan replied:
I don't get it. What's the problem?

rofinn replied:
My point is just that you're going to need to allocate the vector to do anything useful with the weights anyways, so why bother special casing them just because you could compute the weights on the fly.

nalimilan replied:
Well at least you'll avoid storing a full vector all the time, and you save one copy. Also it's not really special casing: exponential weights are already special since they can be computed programmatically, contrary to other weights which are arbitrary. They could be complemented with unit weights, weights that sum to one or other kinds of weights in the future.

rofinn replied:
A new reason to store all the vector might be if we want exponential weights computed based on timestamps. For example)

```
eweights(time::Vector{T}, range::AbstractRange{T}, λ) where T<:AbstractDateTime
```

In this case, the `time` vector represents a set of observation timestamps relative to a datetime `range`. If we have missing observations `time`s in that range then we want to return weights which reflect that.

rofinn replied:
On a separate note, is the `AbstractDateTime` method is something we might want to include here?

nalimilan replied:
Maybe. How would you define that method?

However, exponential weights more and more appear as a different kind than other weights. The main goal of supporting the current three weights types is to apply corrections automatically (e.g. for `var`). That's not the case for exponential weights since no correction applies (https://github.com/JuliaStats/StatsBase.jl/pull/401#discussion_r214542438). So `eweights` just seems like a convenience function to generate a `Vector{Float64}`, and I'm not sure there's any reason to define a new `AbstractWeight` type, especially since https://github.com/JuliaLang/julia/pull/31395 makes it unnecessary for dispatch by using a keyword argument.

rofinn replied:
> However, exponential weights more and more appear as a different kind than other weights.

Agreed. I often think that having a concept such as "importance" weights would be useful. I think STATA already has this?

> I'm not sure there's any reason to define a new AbstractWeight type, especially since JuliaLang/julia#31395 makes it unnecessary for dispatch by using a keyword argument.

I think it's still helpful to store the precomputed sum and it's useful for distinguishing weights from other array types even if base switches to using a keyword weights argument.

nalimilan replied:
As I noted in a previous thread, importance weights in Stata are just arbitrary weights, i.e. equivalent to our `Weights`.

rofinn replied:
Okay, sorry, it's been a while. Would you be okay if we changed this PR to have `eweights` return `Weights`?

nalimilan replied:
Yes, in that case it would make more sense to allocate a `Vector{Float64}`, since you would be allowed to modify values without creating a contradiction with the name of the type.

rofinn replied:
Alright, I think the changes I've made provide a more general/flexible `eweights` method, without the bias correction assumptions.

ararslan commented on src/weights.jl:
```diff
@@ -193,9 +195,87 @@ pweights(vs::RealArray) = ProbabilityWeights(vec(vs))
+"""
+@inline function varcorrection(w::ExponentialWeights, corrected::Bool=false)
+    if corrected
+        1 / (1 - sum(abs2, w.values))
```
It comes up as the correction factor for computing the variance of a weighted mean, but aside from that I don't know.

nalimilan replied:
Yes, but only of the analytical-weighted mean. And I still don't see a reason to think exponential weights should be considered as analytical weights.

ararslan replied:
I don't know what other kind of weights they would be if not analytic. If you have an exponentially weighted moving average, aren't you weighting successive aggregations, and thus using analytic weights?

nalimilan replied:
I don't think so. Analytic weights follow a very specific definition: they represent the number of observations that have been averaged, when giving them equal weights. This determines the variance of the estimator. With exponential weights, the weight does not represent the number of observations which have been averaged, but the weight you give to a certain observation in the computation of the average.

Anyway we need a proof or a reference. We can't just define a method because it might be correct.

rofinn replied:
Hmmm, it's possible that these are more like importance weights in stata, but I'm unclear from the documentation what kind of bias correction, if any, is applied to those weights.

nalimilan replied:
None. "Importance weights" is just the Stata name for "we don't know what these weights mean". So no correction can be applied.

ararslan commented:
I've pared this down to remove the automatic normalization and I've deleted the variance correction method (so exponential weights will fall back to the generic definition which produces an error), corrected the language used in the documentation, removed the `eweights` method that accepts a vector, and removed the tests for functionality no longer implemented.

omus commented:
Bump

nalimilan commented:
Still need to address https://github.com/JuliaStats/StatsBase.jl/pull/401/files#r214542378

rofinn commented:
@ararslan Are you okay if I update your branch?

ararslan commented:
Yes, please do!

rofinn commented on src/weights.jl:
````diff
@@ -193,6 +193,86 @@ pweights(vs::RealArray) = ProbabilityWeights(vec(vs))
+ 0.6122448979591837
+ 1.249479383590171
+```
+"""
````
I'm not sold on this docstring as it's getting pretty long now. Maybe I should move the examples into the markdown docs?

nalimilan commented on docs/src/weights.md:
````diff
@@ -50,6 +50,60 @@ w = Weights([1., 2., 3.])
 w = weights([1., 2., 3.])
 ```
 
+### `eweights`
````
```suggestion
### Exponential weights: `eweights`
```

nalimilan commented on docs/src/weights.md:
```diff
@@ -50,6 +50,60 @@ w = Weights([1., 2., 3.])
+Exponential weights are a common form of temporal weights which assign exponentially decreasing
+weight to past observations.
+
+For each element `i` in `t` the weight value is computed as:
```
What is `t`?

rofinn replied:
`t` is just a vector of arbitrary time indices `i` used in the equation below. The reason for this is that it doesn't need to assume ordering or continuity.

nalimilan replied:
What I meant is that this should be explained to the reader.

nalimilan commented on docs/src/weights.md:
```diff
@@ -50,6 +50,60 @@ w = Weights([1., 2., 3.])
+
+``λ (1 - λ)^{1 - i}``
+
+``λ`` is a smoothing factor or rate parameter such that ``0 < λ \\leq 1``.
```
```suggestion
``λ`` is a smoothing factor or rate parameter such that ``0 < λ ≤ 1``.
```

nalimilan commented on docs/src/weights.md:
````diff
@@ -50,6 +50,60 @@ w = Weights([1., 2., 3.])
+ 7.434279666747019
+```
+
+Finally, passing arbitrary times and a step range is equivalent to passing `something.(indexin(t, r))`.
````
Can you explain using words what this code means?

rofinn replied:
For each value in `t` return the corresponding index for that value in `r`. `indexin` returns `nothing` if there is no corresponding value from `t` in `r`, but that doesn't really make sense in this case.

nalimilan replied:
I mean, add these words to the manual. :-D

nalimilan commented on docs/src/weights.md:
```diff
@@ -50,6 +50,60 @@ w = Weights([1., 2., 3.])
+### `eweights`
+
+Exponential weights are a common form of temporal weights which assign exponentially decreasing
+weight to past observations.
```
```suggestion
weights to past observations.
```

nalimilan commented on src/weights.jl:
```diff
@@ -193,6 +193,60 @@ pweights(vs::RealArray) = ProbabilityWeights(vec(vs))
+
+``λ (1 - λ)^{1 - i}``
+
+``λ`` is a smoothing factor or rate parameter such that ``0 < λ \\leq 1``.
```
```suggestion
``λ`` is a smoothing factor or rate parameter such that ``0 < λ ≤ 1``.
```

nalimilan commented on src/weights.jl:
```diff
@@ -193,6 +193,60 @@ pweights(vs::RealArray) = ProbabilityWeights(vec(vs))
 end
 
+"""
+    eweights(t::AbstractVector{<:Integer}, λ::Real)
```
Why restrict this to integers, while the next method accepts any `T`? Shouldn't `Real` be used in both cases?

rofinn replied:
In this method each element of `t` represents a index in time. In the method below we use `T` because we're generating the weights from the indices of `t` in `r`.

nalimilan replied:
OK. Shouldn't the docstring mention `r` somewhere?

nalimilan commented on src/weights.jl:
```diff
@@ -193,6 +193,60 @@ pweights(vs::RealArray) = ProbabilityWeights(vec(vs))
+    end
+
+    s = sum(w0)
+    Weights{typeof(s), eltype(w0), typeof(w0)}(w0, s)
```
Doesn't this work?
```suggestion
    Weights(w0, s)
```

rofinn replied:
Oops, yep.

nalimilan commented on src/weights.jl:
```diff
@@ -193,6 +193,60 @@ pweights(vs::RealArray) = ProbabilityWeights(vec(vs))
+end
+
+function eweights(n::Integer, λ::Real)
+    n > 0 || throw(ArgumentError("cannot construct exponential weights of length < 1"))
```
Why isn't this checked in the method above too?

rofinn replied:
Honestly, I'm not sure there's a good reason to keep this check. It wouldn't be very helpful, but I don't see a reason we couldn't support empty exponential weights?

nalimilan commented on test/weights.jl:
```diff
@@ -2,6 +2,8 @@ using StatsBase
 using LinearAlgebra, Random, SparseArrays, Test
 
 @testset "StatsBase.Weights" begin
+# NOTE: Do not add eweights here, as its methods don't match those of the others, so the
```
I think this comment isn't needed: it's clear `eweights` is different, if only because it doesn't have a corresponding type.

rofinn replied:
Okay

rofinn commented:
Okay, apart from your comments around the `eweights(t, r, λ::Real)` I think that's everything.

nalimilan commented on src/weights.jl:
```diff
@@ -193,6 +193,61 @@ pweights(vs::RealArray) = ProbabilityWeights(vec(vs))
+    eweights(t::AbstractVector{T}, r::StepRange{T}, λ::Real) where T
+    eweights(n::Integer, λ::Real)
+
+Construct [`Weights`](@ref) vector which assigns exponentially decreasing weights to past
```
```suggestion
Construct a [`Weights`](@ref) vector which assigns exponentially decreasing weights to past
```

nalimilan commented on src/weights.jl:
```diff
@@ -193,6 +193,61 @@ pweights(vs::RealArray) = ProbabilityWeights(vec(vs))
+
+Construct [`Weights`](@ref) vector which assigns exponentially decreasing weights to past
+observations, which in this case corresponds to larger integer values `i` in `t`.
+
```
`n` should be mentioned here I think. I suggest something like:
```suggestion
If an integer `n` is provided, weights are generated for values from 1 to `n`
(equivalent to `t = 1:n`).

```

nalimilan commented on src/weights.jl:
```diff
@@ -193,6 +193,61 @@ pweights(vs::RealArray) = ProbabilityWeights(vec(vs))
+
+- `t::AbstractVector`: temporal indices or timestamps
+- `r::StepRange`: a larger range to use when constructing weights from a subset of timestamps
+- `n::Integer`: number of temporal indices to assume for the exponential weighting
```
```suggestion
- `n::Integer`: if provided instead of `t`, temporal indices are taken to be `1:n`
```

rofinn commented:
I guess if nobody has any other comments then I'll merge it tomorrow.

ararslan commented:
Thanks for getting this finished up, Rory. :)
